sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/AdaptiveSparkPlanExec.scala

```diff
@@ -91,13 +91,22 @@ case class AdaptiveSparkPlanExec(
     DisableUnnecessaryBucketedScan
   ) ++ context.session.sessionState.queryStagePrepRules
 
```

Comment:
we can make `queryStagePreparationRules` to take a parameter `optimzeSkewJoin: Boolean`, and call it twice in `reOptimize`, to produce 2 plans. Then I don't think we need the hack for the final stage anymore?